### ISSUE TITLE
History Graph in comparison view

### ISSUE THREAD

**Issue by hottestchilipepper** [desktop/desktop#9452]
### Describe the feature or problem you’d like to solve

I need a Graph feature. I know this issue is duplicated with #1634 , but it is locked, and @outofambit  allowed me to reopen this issue at #9444 . I implemented but I also think it needs more discussion.

Sometimes I have to know where the branch comes from and when it was made. But I forget often where the branch comes from, and I have to look for history reading so many commit messages. I think this is not just for me. Human cannot remember what all branches come from.

Merge direction is important too. Assume that I made a 'branch-a' from 'master', after some commits were pushed on 'master' and 'branch-a', made a 'branch-b' from 'branch-a'. I should merge 'branch-a' to 'branch-b' and 'master' to 'branch-a'. I should not merge 'master' to 'branch-b' directly and should not merge 'branch-b' to 'branch-a'. In order to know the merge direction, we have to make documents about all branches. But it's annoying, it will be better to see from History Graph. And we can see intuitively from History Graph whether juniors made mistakes in merging.

As many people said, in order to aware that current branch is out of date, History Graph is also useful, too.

### Proposed solution

The full graph is very complicated. Following the purpose of it, it will be better to show graph between only two branches. Many people want the classic thing. I mixed these needs with modern GitHub Desktop. So I propose below

![K-042](https://user-images.githubusercontent.com/7362974/78418811-d6335f00-767a-11ea-8b7e-5e4af0608910.png)

History Graph in Comparison View. With this, we can see whether current branch comes from comparison branch. Also we can see the merge direction and when the current branch was made.

Any questions, ideas and advice will be helpful.

**Comment by keyboard-clacker** [comment]
++ To this! One other use case I currently have for some sort of graph view is that my team uses trunk-based development, and it would be incredibly useful to be able to search through a graph and see the different commits one after another similar to navigating through commits in a PR

**Comment by Mlocik97-issues** [comment]
I would much more like to have that graph horizontally, as in old Github 1.x

**Comment by TheJaredWilcurt** [comment]
My concern with this design is that it may be "good enough" and prevent adoption of the horizontal graph used previously.

![](https://camo.githubusercontent.com/a6a00dd0c59a72cfda872e9825eb61ae1504c26e/68747470733a2f2f676974687562656e67696e656572696e672e636f6d2f696d616765732f63726f73732d706c6174666f726d2d75692d696e2d6769746875622d6465736b746f702f636f6d70617269736f6e2d67726170682d6f732d782e706e67)

A vertical graph is basically all you can do in a terminal window, you can only scroll up and down. But a horizontal graph is much friendlier and easier to read at a glance. I do like that in the proposed vertical design that it limits the graph to two branches. I don't think I've ever wanted to compare more branches than just my current and one other (99% of the time it's the master/main/default branch).

Right now the only horizontal graph I'm aware of is in Classic GH Desktop (very hard to procure) and gmaster (overly complicated).

**Comment by bhadresharya** [comment]
I'm also having same requirement. I have to open VS Code and Git Graph extension every time to check and track the branches.
Vertical Graph something like this would be so useful. Vertical graph will be more easy to visualize all commits with dots and merging with lines. 

![image](https://user-images.githubusercontent.com/6179421/102745389-dea30580-4381-11eb-94d5-4e829e5017ee.png)

**Comment by whippingdot** [comment]
Wiil this ever happen? When will the graph be implemented?

**Comment by MrHackino** [comment]
I use GitAhead and VSCodium because GH Desktop doesn't have this feature. 
It would be nice to have a graph and it may makes me switch.

**Comment by TheJaredWilcurt** [comment]
There's like 40 apps out there to visualize vertical graphs, with way too many branches ("Fork" being one of the better ones if you're into that kind of thing). There's almost none with a simple horizontal graph. Don't give us "yet another" of the same thing we already have plenty of. Bring back your best feature.

**Comment by yoseph1998** [comment]
At least bring it back and make it optional to enable.

**Comment by trafium** [comment]
I am a simple guy with simple requirements towards my git client. I was using old native Github Desktop client for a long time until it no longer worked since a couple months ago. I tried new version  a while ago and recently gave it another try, but I must say I found it to be a huge downgrade because of this missing feature. Even after weeks of attempts I felt half-blinded working with git without any visual aid and had to switch to another client (with huge multibranch graph, which is overwhelming, but god, it's better than nothing).

**Comment by Echsecutor** [comment]
This might be a reopen of https://github.com/desktop/desktop/issues/1634 (sorry for that) but showing the history (all of it, not only the linearised past of the current branch head) is THE essential feature of a git GUI like gitk https://git-scm.com/docs/gitk . All of those do it: https://git-scm.com/downloads/guis , the competitor gitlab does it in the web UI https://docs.gitlab.com/ee/user/project/repository/#repository-graph ... that is for a good reason.

Seeing the graph is absolutely essential for working on a git repository with more than 2 people/branches at the same time in order to understand what is going on, i.e. where branches took off and what is merged in where. Please add a git GUI to the github web UI and the desktop!

**Comment by TheJaredWilcurt** [comment]
GitHub already has that feature under insights > network. It's useful for seeing the network of forked branches:

* **Example:** https://github.com/nwjs/nw.js/network

If you want that locally, use one of the other existing UI's that already offer this. GitHub Desktop should focus on helping people get started with Git, and in simplifying the complexities of Git. The previous [2-track horizontal git graph](https://camo.githubusercontent.com/a6a00dd0c59a72cfda872e9825eb61ae1504c26e/68747470733a2f2f676974687562656e67696e656572696e672e636f6d2f696d616765732f63726f73732d706c6174666f726d2d75692d696e2d6769746875622d6465736b746f702f636f6d70617269736f6e2d67726170682d6f732d782e706e67) worked very well for this. Vertical graphing is available in the command line and in every other UI. Showing all branches is confusing and overwhelming. It would contain irrelevant information to what you care about. Just show 2 branches, horizontal. Keep it simple, if people want complex, that's what everyone else is offering.

**Comment by Echsecutor** [comment]
Many thanks for pointing out the Network Graph Feature! Not the best GUI view, but definitely essential. Seeing a git history like this is, in my view, essential to understand what git does in the first place, even for new commers.

**Comment by whippingdot** [comment]
To GitHub(P.S. this is taken from something I wrote in an issue called Warning):

I have had an idea for adding the Git Graph. If you are hesitant of adding a Git Graph because there is no space in the application, you can add a button in the top row as there is a lot of white space:

**ON A BIG COMPUTER LIKE MINE**
![image](https://user-images.githubusercontent.com/64354499/112972726-31087500-916e-11eb-939a-d25a9e845021.png)

**REDUCED FRAME**
![image](https://user-images.githubusercontent.com/64354499/112972832-51d0ca80-916e-11eb-8f78-5d2da8797791.png)

When you press the button, it can take you to another tab with a whole scrollable git graph with any branch's name (even deleted ones) on hover and full commit details.

**VS CODE GIT GRAPH COMMIT DETAILS**
![image](https://user-images.githubusercontent.com/64354499/112973295-c99ef500-916e-11eb-9318-2c01bec28169.png)

_Note: This is how the full commit details can look, though I would prefer them on a popup tab below or with a better style. This doesn't look that nice so if you implement this, I would expect a better design_

**FULL GIT GRAPH LOOK**
![image](https://user-images.githubusercontent.com/64354499/112973624-30241300-916f-11eb-9c74-6072a158517d.png)

_Note: Picture taken from online - [Axosoft](http://blog.axosoft.com/wp-content/uploads/2015/10/1-main-view.png). This is a picture of GitKraken. Again I would expect a better design from GitHub Desktop but this looks sweet anyway and I love that the colors pop out!_

**Comment by JenuelDev** [comment]
Any update on github desktop git graph?

**Comment by goyalyashpal** [comment]
> a horizontal graph is much friendlier and easier to read at a glance
_TheJaredWilcurt at https://github.com/desktop/desktop/issues/9452#issuecomment-698104410_

yeah, this point was also raised at https://github.com/desktop/desktop/issues/1634#issuecomment-303646773

**Comment by goyalyashpal** [comment]
> https://docs.gitlab.com/ee/user/project/repository/#repository-graph ... that is for a good reason.
@Echsecutor at https://github.com/desktop/desktop/issues/9452#issuecomment-779948669

the link needs update: https://docs.gitlab.com/ee/user/project/repository/#repository-history-graph

**Comment by JenuelDev** [comment]
does git graph already in github desktop? that would be nise.

**Comment by niclasmattsson** [comment]
I think @hottestchilipepper's concept is brilliant. Let me explain why since many here don't seem to appreciate it:

* It looks a bit bland right now, but once the professional UI designers in the Desktop team add a tiny bit of color and tweak a few other graphical elements I think it will be gorgeous.
* Limiting the graph to the comparison view is the key insight here. I'm convinced that the reason why we still don't have a graphical branch view in Desktop yet is because the dev team  thinks that the "rainbow spaghetti" charts (see examples above) in all other git GUIs are confusing and conflict with the barebones simplicity of Desktop. I agree. A simple two-branch view is a perfect compromise.
* To those pushing for a horizontal graph: I agree that the horizontal graph looks clean and simple. But the problem with that design is that it adds no information to the "X commits ahead, Y commits behind" text that Desktop already presents. The advantage of a vertical graph is that you can show the actual commits side-by-side (as in this suggestion) which adds **context** to the ahead/behind text, making it a valuable addition to the comparison view.

In short: this is an excellent proposal. Now that Desktop finally has rollback functionality (YAY to "Create branch from commit") this is the last thing missing for Desktop to become "feature complete" for day-to-day git management - and IMHO the best git GUI available. I sincerely hope @billygriffin and the rest of the dev team revisit the idea and start hacking on it.

**Comment by goyalyashpal** [comment]
> * .... The advantage of a vertical graph is that you can show the actual commits side-by-side (as in this suggestion) which adds **context** to the ahead/behind text, making it a valuable addition to the comparison view.

yeah, i was thinking that yesterday. Horizontal graph looses the ability to use the vertical dimension and also the efficiency in  visually linking with the info in commit history - when language is horizontal as well (which most languages are).

**Comment by JenuelDev** [comment]
this image is taken from VS Code git graph extension... 
if your a vscode user,, u can use this extension instead..
![image](https://user-images.githubusercontent.com/31676163/122870504-51f0f880-d360-11eb-8f59-83ce1dba38b2.png)

**Comment by ssafiat** [comment]
Envoyé de mon iPhone

**Comment by whippingdot** [comment]
> this image is taken from VS Code git graph extension...
> if your a vscode user,, u can use this extension instead..
> ![image](https://user-images.githubusercontent.com/31676163/122870504-51f0f880-d360-11eb-8f59-83ce1dba38b2.png)

We know that, but we want it implemented in GitHub Desktop. Not everyone uses VSCode, and this is a thread about implementing the feature in GitHub Desktop so...

**Comment by a-szotyori** [comment]
Just my 2 cents: I wanted to switch from Sourcetree to GitHub desktop because we have so many repos there, but the lack of a history graph is a big issue for me. It gives me a quick overview of the history and interplay between branches. Just look at how much useful information I get in Sourcetree in one screen:
<img width="1555" alt="Screenshot 2021-09-23 at 11 43 55" src="https://user-images.githubusercontent.com/13149498/134487124-5454db82-a2c5-4600-a365-f7425b92cc0f.png">

The history graph is prominent to provide easy navigation and "situational awareness". There are also options to filter and customise the graph. For each selected commit I can easily browse the changes in the bottom of the screen.

Please reconsider this feature request. Until then I'll stick with Sourcetree.

**Comment by niclasmattsson** [comment]
No offense, but I don't think those screenshots from VScode and Sourcetree are a good fit for Github Desktop, which is all about simplicity, clarity and minimalism. I called graphs like that "rainbow spaghetti" in my comment above, and I honestly think those examples are more deterring than encouraging to the core Desktop devs. In contrast, I think the simple 2-branch graph in Desktop's comparison view as proposed in the OP is a more realistic idea and a perfect fit for Desktop.

**Comment by a-szotyori** [comment]
Thanks @niclasmattsson for your comment. I understand now that the design philosophy is different for GitHub Desktop. There is no one-size-fits-all and I discovered now that GitHub Desktop is not for me.

P.S.: I would recommend refraining from using somewhat derogatory terms that you invented publicly with strangers. It might be "rainbow spaghetti" for your and for someone else it's a valuable productivity booster.

**Comment by gh-jvirdee** [comment]
is there anything WIP for this ticket?

**Comment by JenuelDev** [comment]
is History Graph in Github desktop, will it be implemented?

**Comment by vedrancu** [comment]
Hi,
history graph in Sourcetree is very usefull tool. It's pity that such feature is not available in GiHub Desktop. I must go to Sourcetree app.

**Comment by vrgimael** [comment]
Seems like there's this attempt? Anybody knows why was this PR closed and never merged?

https://github.com/desktop/desktop/pull/9444

**Comment by SpaceShot** [comment]
A lack of a way to visualize the commit graph is the single reason I recommend to others not to use GitHub Desktop.  I understand all the arguments for simplicity, but I find GitHub Desktop users merge from main more and then wonder why code "they thought they fixed" has returned.  They often are the users who say that Git forces you to save off your changes elsewhere, dump and reclone your repo, and copy them back in.  It seems really counterproductive.

I have had some success with SourceTree, but the tool I've been sticking with is the one that lets me hide all the remote branches and even my own so I can focus on what's important.  I think even those tools could be improved and smarter, letting me say I'm not interested in seeing tags, not interested in seeing new remote branches, just let me focus and narrow my view to what's important.  That alone has allowed me to be very productive, but of course this is more than a "starter" feature.  

One must first be very comfortable with what a branch/tag pointer really are in Git.   I know that sounds like I am advocating for everyone to be advanced, but I've found one I teach someone this one basic point about Git, they seem to become much more productive and less afraid since they know a branch is just a pointer to a commit in a permanent graph.

**Comment by mananrrg95** [comment]
Saya bukan pr saya warga negare saya igin off semua...

Pada Jum, 4 Mac 2022 21:34 Chris Gomez ***@***.***> menulis:

**Comment by Remusqs1** [comment]
This is a critical feature and is very confusing why the official destop app of github lacks of a graphical tree view. This is the only reason why I have to use other's git clients to work with.

Please add this feature

**Comment by pebf** [comment]
I don't want to get back to slow and heavy git browsers. But due to Github Desktop's lack of graph, I have to do it. Bye and see you when you have it.

**Comment by TheJaredWilcurt** [comment]
![](https://camo.githubusercontent.com/a6a00dd0c59a72cfda872e9825eb61ae1504c26e/68747470733a2f2f676974687562656e67696e656572696e672e636f6d2f696d616765732f63726f73732d706c6174666f726d2d75692d696e2d6769746875622d6465736b746f702f636f6d70617269736f6e2d67726170682d6f732d782e706e67)


Co-worker yesterday saw me using GitHub desktop for Windows from like 2016 (pre-electron), and was like "wow that's awesome, what is that? It looks sort of like GitHub desktop... but better". I had to explain that it was both GitHub desktop, *and* better, and then he was sad that he couldn't get it anymore.

* * *

**Edit:** I can say this happens multiple times per year now

**Comment by JenuelDev** [comment]
any update on this?

**Comment by piotrku** [comment]
Yeah this is one of those must have features but the developers of this project for years seem to be more more interested in other things.

Having an overview is really important in bigger projects - it's then easy to see what everyone is/was doing and where in just a moment. There's nothing good in simplicity when you cannot be as productive as you could be with other tools.

**Comment by ozgurrgul** [comment]
It's weird that GH Desktop doesn't have one of the most important features.

**Comment by erano067** [comment]
Hi, I think I can take that mission, my question is if I will have a branch in a forked repo and work on that graph feature with your opinions from this thread and my thoughts of how it should looks, will it be considered to add? or it is not in the roadmap so my work will be thrown away?

**Comment by Benoit1980** [comment]
This is a super important feature, sad :-(

**Comment by MinhCuongIT** [comment]
I also really need this feature for my work

**Comment by Benoit1980** [comment]
> I also really need this feature for my work

What I did for now is also install sourcetree in parallel and use the graph of sourcetree but GIT with GitDesktop as I find it easier to use. Not sure if this will help you. But if they do not add this feature in Git Desktop, I will have to drop it, sad because it is such great little piece of software.

**Comment by aisrael-fss** [comment]
Adding another voice to this request. Please add a history graph.

**Comment by philip-nicholls** [comment]
@niclasmattsson If you think that is "rainbow spaghetti" you've clearly not used feature branches in git very much, which is a huge part of what git is used for.  You can of course choose to avoid using feature branches in your repository to keep things simple with just 1 or 2 branches, but most people have (forgive the pun) "branched out" and really need this graph feature.  I also believe your intentions are good, but your philosophy is ultimately flawed.   You are trying to hide the complexity of git so it looks simple to a novice user, but it doesn't make it simpler under the hood and it annoys seasoned developers who use and expect such a basic feature.

**Comment by ARBS23** [comment]
Please add a history graph.

**Comment by lfaveur** [comment]
This will be good have it inside github desktop to avoid us use the website github to consult the branches or use another app like sourcetree or gitkraken ... it will be a complete tool.

**Comment by somehibs** [comment]
GitHub has commit graphs, but many companies and individuals use self-hosted git solutions with GitHub Desktop for various reasons (legal constraints, privacy, etc).

Users of self-hosted git solutions would benefit from commit graphs. I have had two colleagues tell me that this missing feature prevents them from switching, so they would stick with SourceTree.

**Comment by Blosberg** [comment]
re: "keeping things simple"

There's a big difference between 
(a) "keeping things simple" (by removing irrelevant or unimportant information),
and 
(b) _oversimplifying_ things: making them _appear_ to be simpler than they actually are by hiding relevant and important information that is actually needed.

If a project is indeed simple (having only 1 or 2 branches) then this whole point is moot, but for situations where the project graph structure is more complex, then making it impossible to _see_ that complexity  is an example of (b). Where my current branch is in a project is the most important piece of contextual information I want in repo management.

re "other packages offer this"

The reason everyone else includes this is because _it's a crucial piece of functionality_.  Without the graph, Github Desktop is like a car without a windshield -the engine is powerful and the seats are comfy, but if I can't see where I am and where I'm going, then there's no situation in which I'm going to feel safe driving it anywhere.

**Comment by vedrancu** [comment]
More than on year we are talking about graph and still not added. Sad.  :(

**Comment by kinghuang** [comment]
> More than on year we are talking about graph and still not added. Sad. :(

I wish it was only one year! We've been talking about this for over 5 years. The [original issue](https://github.com/desktop/desktop/issues/1634) was opened in May 2017.

**Comment by Munna231** [comment]
> ### Describe the feature or problem you’d like to solve
> I need a Graph feature. I know this issue is duplicated with #1634 , but it is locked, and @outofambit allowed me to reopen this issue at #9444 . I implemented but I also think it needs more discussion.
> 
> Sometimes I have to know where the branch comes from and when it was made. But I forget often where the branch comes from, and I have to look for history reading so many commit messages. I think this is not just for me. Human cannot remember what all branches come from.
> 
> Merge direction is important too. Assume that I made a 'branch-a' from 'master', after some commits were pushed on 'master' and 'branch-a', made a 'branch-b' from 'branch-a'. I should merge 'branch-a' to 'branch-b' and 'master' to 'branch-a'. I should not merge 'master' to 'branch-b' directly and should not merge 'branch-b' to 'branch-a'. In order to know the merge direction, we have to make documents about all branches. But it's annoying, it will be better to see from History Graph. And we can see intuitively from History Graph whether juniors made mistakes in merging.
> 
> As many people said, in order to aware that current branch is out of date, History Graph is also useful, too.
> 
> ### Proposed solution
> The full graph is very complicated. Following the purpose of it, it will be better to show graph between only two branches. Many people want the classic thing. I mixed these needs with modern GitHub Desktop. So I propose below
> 
> ![K-042](https://user-images.githubusercontent.com/7362974/78418811-d6335f00-767a-11ea-8b7e-5e4af0608910.png)
> 
> History Graph in Comparison View. With this, we can see whether current branch comes from comparison branch. Also we can see the merge direction and when the current branch was made.
> 
> Any questions, ideas and advice will be helpful.

**Comment by Benoit1980** [comment]
> > ### Describe the feature or problem you’d like to solve
> > I need a Graph feature. I know this issue is duplicated with #1634 , but it is locked, and @outofambit allowed me to reopen this issue at #9444 . I implemented but I also think it needs more discussion.
> > Sometimes I have to know where the branch comes from and when it was made. But I forget often where the branch comes from, and I have to look for history reading so many commit messages. I think this is not just for me. Human cannot remember what all branches come from.
> > Merge direction is important too. Assume that I made a 'branch-a' from 'master', after some commits were pushed on 'master' and 'branch-a', made a 'branch-b' from 'branch-a'. I should merge 'branch-a' to 'branch-b' and 'master' to 'branch-a'. I should not merge 'master' to 'branch-b' directly and should not merge 'branch-b' to 'branch-a'. In order to know the merge direction, we have to make documents about all branches. But it's annoying, it will be better to see from History Graph. And we can see intuitively from History Graph whether juniors made mistakes in merging.
> > As many people said, in order to aware that current branch is out of date, History Graph is also useful, too.
> > ### Proposed solution
> > The full graph is very complicated. Following the purpose of it, it will be better to show graph between only two branches. Many people want the classic thing. I mixed these needs with modern GitHub Desktop. So I propose below
> > ![K-042](https://user-images.githubusercontent.com/7362974/78418811-d6335f00-767a-11ea-8b7e-5e4af0608910.png)
> > History Graph in Comparison View. With this, we can see whether current branch comes from comparison branch. Also we can see the merge direction and when the current branch was made.
> > Any questions, ideas and advice will be helpful.

In my case, I always work with a minimum of 3 branches, the main, the staging/develop and the update(min of 1 but most likely 2). 2 Branches might be a bit too little, I think. Min of 4 would be nice.

**Comment by NirmalScaria** [comment]
**ONLY FOR GITHUB WEB**

If someone is looking for a solution/workaround to use git graph for GitHub, you could use Le Git Graph, a browser extension that does exactly this.

[![Le git graph demo][1]][1]


Install the extension from here : [https://chrome.google.com/webstore/detail/le-git-graph-commits-grap/joggkdfebigddmaagckekihhfncdobff][2]

It will add a new "commits" section to every GitHub repo you open. Open the commits graph and there, all commits across branches will be listed along with the git graph.


[![Le git graph demo 2][3]][3]

Hope it helps! 

  [1]: https://i.stack.imgur.com/jLC3u.jpg
  [2]: https://chrome.google.com/webstore/detail/le-git-graph-commits-grap/joggkdfebigddmaagckekihhfncdobff
  [3]: https://i.stack.imgur.com/tesWi.jpg

**Comment by Gakk** [comment]
@NirmalScaria The extension _Le Git Graph_ has a beautiful look, and gives exactly what is needed.

Unfortunately I do not see it fit for private repositories. For public repositories it should work great, even though I have not tested it.

If it had been rewritten to accept a personal access token (PAT) _(e.g. the same way as [Refined GitHub](https://github.com/refined-github/refined-github) does)_, a fine-grained token could have been supplied, but authenticate directly (through your server) as a user is unfortunately a big no-go for me.

_Feature request for Le Git Graph: https://github.com/NirmalScaria/le-git-graph/issues/33_

**Comment by cloud8mike** [comment]
Any news on this? This has got to me the most desired feature. 

I would love to see this feature added and is the only keeping me from using Github Desktop as my main client. I usually work across many branches, so although a visual graph between two would be useful, it would not be ideal. I find Sublime Merge does this the best where I can see everything at a glance. It does get to be _a lot_ when there are many branches, but if Github Desktop could switch between this and just current/default branch that would be the dream.

![image](https://user-images.githubusercontent.com/103044065/221234694-9a6dbb04-3d60-4c71-8b4c-4cb90bcab9f0.png)

**Comment by thomaschampagne** [comment]
As [projektorius96](https://github.com/projektorius96) mentioned in his duplicate ticket (#16186), Mermaid.js could be considered as base for this feature.

You may give a look a this: https://mermaid.js.org/syntax/gitgraph.html

This js lib seems also interesting: https://github.com/nicoespeon/gitgraph.js as base

**Comment by glewe** [comment]
+1
I'd love to see this feature too in Github Desktop

**Comment by Benoit1980** [comment]
+1
Me too!!! :)

**Comment by gasperphoenix** [comment]
Me too :-)

**Comment by Deus-nsf** [comment]
A lot of cool things suggested, but I'll stand by the idea of something horizontal, and that only compares 2 selected branches for better readability. Basically the original design with the same icons, which was just great.

**Comment by BroJenuel** [comment]
the developers of the git graph extension for VS code should donate their talents on this.

**Comment by phil294** [comment]
> the developers of the git graph extension for VS code should donate their talents on this.

I don't think it's appropriate to ask open source volunteers to "donate" to a trillion dollar company.

**Comment by giosueDelgado** [comment]
There are updates here?

**Comment by ttnbtfy** [comment]
Where is my Graph?

**Comment by JenuelDev** [comment]
any update?

**Comment by FrancoisGrosbellet** [comment]
I would love this feature to be available in GH Dekstop too.

**Comment by Deus-nsf** [comment]
I believe we can sit on it :(

**Comment by datfit09** [comment]
+1
Me too!!! :)

**Comment by giosueDelgado** [comment]
There are some updates about this? @niik

**Comment by me-kell** [comment]
It has often been argued that GitHub Desktop is about simplicity.

But branching in git is one of the most important features and a not a trivial one. Branches are a complex issue.

IMHO hiding branches visualization is not keeping it simple at all but avoiding to understand what is going on with branches.

If GitHub Desktop wants to contribute to git's acceptance among new users (specially beginners) it should reconsider adding a branch visualization. (Which could be disabled as default).

**Comment by Deus-nsf** [comment]
I have abandonned this app a long time ago, horizontal clean visualization of branches diffs was it's most important feature and thanks to that I was able to learn git very efficiently, I was lucky to have this opportunity back in the days.

**Comment by notrealdev** [comment]
Is there any graphing software available for free ( not Sourcetree )?
I am weary of awaiting an unresponsive developer to implement this feature.

**Comment by FrancoisGrosbellet** [comment]
[GitKraken](https://www.gitkraken.com) provides a good implementation of this feature.

Several limitations applies to the Free version though, so you would need to ensure it fit your case.

**Comment by bulutfatih** [comment]
> Is there any graphing software available for free ( not Sourcetree )? I am weary of awaiting an unresponsive developer to implement this feature.

You can use [Fork](https://fork.dev/), which is totally free. I find it much better than GitHub.

**Comment by me-kell** [comment]
@bulutfatih AFAICS Fork is not free. See [Fork license page](https://fork.dev/license):

> You can download and evaluate the Software for free, but need to purchase a license for long-term use.

**Comment by notrealdev** [comment]
Fork offers a superior interface compared to GitKraken, however, it is not available for free 🤣

**Comment by giosueDelgado** [comment]
It's incredible that for a feature needed for almost 4 years, from many users, it has not been implemented yet that is so basic and the competitor of the git client has it... (Fork, Git Graph vscode ext, GitKraken, Gitlab, ecc..)

**Comment by steveward** [comment]
Requested in https://github.com/desktop/desktop/issues/18390.

**Comment by tomasdelima** [comment]
+1 add tree views please!

**Comment by tidy-dev** [comment]
Requested in https://github.com/desktop/desktop/issues/18776

**Comment by danelec-hpm** [comment]
+1 to graph of the branches. Only thing preventing me from switching to GitHub Desktop.

**Comment by adrienbricchi** [comment]
Well, at least that's official.
That `not planned` tag, added without any comment...

I guess we just can move on to something else now.

**Comment by piotrku** [comment]
I'm still not sure what is the problem here with the implementation, it could've been so nice piece of software.
Well then time to leave this place.

**Comment by akhilbetta** [comment]
cant believe Microsoft is sitting on this MOST IMPORTANT feature for 4 years! either they have poor developers or they dont give a $#%&..

**Comment by thomaschampagne** [comment]
I personally abandoned believing in that ticket...

I search/tested other viable open-source solutions for months... In the end, I use these 2 solutions with graphs

1. [SourceGit](https://github.com/sourcegit-scm/sourcegit?tab=readme-ov-file#screenshots) **(unknown & underrated today)**
 ![image](https://github.com/user-attachments/assets/403f39fc-9945-4b9d-9936-bcda88296aab)
![image](https://github.com/user-attachments/assets/aa76f409-11bc-43dd-b0f9-6a0882c1080a)

2. [Lazygit](https://github.com/jesseduffield/lazygit?tab=readme-ov-file#commit-graph)
![image](https://github.com/user-attachments/assets/806e1488-19d4-4d41-bbe2-e9617344c815)

I'll never go back on something paid/closed like sourcetree, smartgit, sublime-merge, gitkraken,..

**Comment by marianolc** [comment]
It's incredible how many people here are saying this feature isn't necessary.

**Comment by NikVladi** [comment]
+1... Oh my god, this still hasn't been done?

**Comment by ttnbtfy** [comment]
I don't believe that after ChatGPT 5 released, Github desktop still has nothing happening with this 🤣

**Comment by Brugman** [comment]
+1

**Comment by agil3st** [comment]
It's still open but there's a `not-planned` tag? what is that mean?

**Comment by danelec-hpm** [comment]
> It's still open but there's a `not-planned` tag? what is that mean?

That they won't do it

**Comment by ACerka-Springbuk** [comment]
Just installed gitdesktop because I'm essentially looking at alternative GUIs... but to be real blunt the lack of a good git history visualization is an absolute game breaker. Without this functionality, I will not use Github's desktop client. Which is honestly really sad because it's got better functionality in some of the other areas that I really enjoy.

**Comment by NikVladi** [comment]
Still not done... Well... We continue to use SourceTree...

**Comment by paulcruse3** [comment]
Would also really appreciate this feature. Gitkraken is my tool until it gets added

**Comment by Altairseven** [comment]
howcome this is not a feature yet..

**Comment by rangga-synergy** [comment]
i need this one so badly.

**Comment by londos** [comment]
Found this [official blog post](https://github.blog/news-insights/cross-platform-ui-in-github-desktop/) from this very day in 2015, which described the horizontal comparison graph as "the centerpiece" and "easily the most sophisticated piece of user interface" of the app.  It also seems like a surprisingly decent roadmap for how it was implemented at the time...
